File: Introduction to Robotics Course/Chapter12/Lesson2/Chapter12_Lesson2.java
import org.ros2.rcljava.RCLJava;
import org.ros2.rcljava.node.Node;
import org.ros2.rcljava.publisher.Publisher;
import org.ros2.rcljava.subscription.Subscription;
import std_msgs.msg.Float64;

public class TopicDemo {
  public static void main(String[] args) {
    RCLJava.rclJavaInit();
    Node node = RCLJava.createNode("java_topic_demo");

    Publisher<Float64> pub =
      node.createPublisher(Float64.class, "/u_cmd");

    Subscription<Float64> sub =
      node.createSubscription(Float64.class, "/u_cmd",
        msg -> System.out.println("u_cmd=" + msg.getData()));

    node.createWallTimer(10, () -> {
      Float64 m = new Float64();
      m.setData(1.0);
      pub.publish(m);
    });

    RCLJava.spin(node);
  }
}
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter12/Lesson2/Chapter12_Lesson2.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64

class SinePublisher(Node):
    def __init__(self):
        super().__init__('sine_publisher')
        self.pub = self.create_publisher(Float64, '/u_cmd', 10)
        self.t = 0.0
        self.timer = self.create_timer(0.01, self.step)  # 100 Hz

    def step(self):
        msg = Float64()
        msg.data = 1.0  # placeholder input
        self.pub.publish(msg)

class UListener(Node):
    def __init__(self):
        super().__init__('u_listener')
        self.sub = self.create_subscription(
            Float64, '/u_cmd', self.cb, 10)

    def cb(self, msg):
        self.get_logger().info(f"u_cmd={msg.data}")

def main():
    rclpy.init()
    pub_node = SinePublisher()
    sub_node = UListener()
    executor = rclpy.executors.MultiExecutor()
    executor.add_node(pub_node)
    executor.add_node(sub_node)
    executor.spin()

if __name__ == "__main__":
    main()
      
